### PR TITLE
 Update architecture image, removed unneeded files, renamed runtime wa…

### PR DESCRIPTION
**Pull request by jeremyharisch**
…tcher repo, include more details in docs (not fully finished

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,19 +3,26 @@
...
+
 
 ## Components
 
 The workflow of Runtime Watcher includes the following main components:
 
-- Watcher custom resources (CRs)
-- Runtime Watcher
-- Listener package
+### [Watcher custom resources (CRs)](https://github.com/kyma-project/lifecycle-manager/blob/main/api/v1beta2/watcher_types.go)
```
Comment (mmitoraj) [inline README.md]:
```suggestion
### Watcher custom resources (CRs)
```

Comment (mmitoraj) [reply README.md]: I would move that link from the header to the paragraph.

File: README.md [hunk 2]
```diff
@@ -3,19 +3,26 @@
...
+The Watcher CRs configures the [Virtual Services](https://istio.io/latest/docs/reference/config/networking/virtual-service/) in the KCP, which are used as a Reverse-Proxy to route incoming request to the correct operator. Furthermore, they are used to configure the [Runtime Watcher](###runtime-watcher) deployed in each Kyma cluster. More details can be found [here](./docs/05-10-api.md).
```
Comment (mmitoraj) [inline README.md]:
```suggestion
[Watcher CRs](https://github.com/kyma-project/lifecycle-manager/blob/main/api/v1beta2/watcher_types.go) configure the [Virtual Services](https://istio.io/latest/docs/reference/config/networking/virtual-service/) in KCP, which are used as a reverse proxy to route incoming requests to the correct operator. Watcher CRs are also used to configure the Runtime Watcher deployed in each Kyma cluster. For more details, see the [Watcher CR](./docs/05-10-api.md) document.
```

File: README.md [hunk 3]
```diff
@@ -3,19 +3,26 @@
...
+
+### [Runtime Watcher](https://github.com/kyma-project/runtime-watcher/tree/main/runtime-watcher)
```
Comment (mmitoraj) [inline README.md]: The link doesn't work. What did you want to link to? I would avoid links in headers and move that one if needed.
```suggestion
### Runtime Watcher
```

File: README.md [hunk 4]
```diff
@@ -3,19 +3,26 @@
 
 ## Overview
 
-Runtime Watcher is a validation webhook deployed by [Lifecycle Manager](https://github.com/kyma-project/lifecycle-manager) on a Kyma cluster. It watches changes on the resources configured in Watcher custom resources (CRs) in Kyma Control Plane (KCP).
+Runtime Watcher is a validation webhook deployed by [Lifecycle Manager](https://github.com/kyma-project/lifecycle-manager) on a Kyma cluster. It watches changes on the resources, configured by Watcher custom resources (CRs) in Kyma Control Plane (KCP).
```
Comment (mmitoraj) [inline README.md]:
```suggestion
Runtime Watcher is mostly a validation webhook deployed by [Lifecycle Manager](https://github.com/kyma-project/lifecycle-manager) in a Kyma cluster. It watches changes in the resources, configured by Watcher custom resources (CRs) in Kyma Control Plane (KCP).
```

File: README.md [hunk 5]
```diff
@@ -3,19 +3,26 @@
...
+
+### [Runtime Watcher](https://github.com/kyma-project/runtime-watcher/tree/main/runtime-watcher)
+The Runtime Watcher consists of multiple parts. First of all, it has a ValidationWebhook which is re-used, instead of its original usecase, for the general watch mechanism inside the SKR. This webhook is configured by the Lifecycle Manager using the Watcher CRs. In addition, a deployment is attached to this webhook, which is the receiver for the validation requests. The deployment converts the validation requests into a [WatchEvent](https://github.com/kyma-project/runtime-watcher/blob/de040bddeba1a7875e3a0e626db4634134971022/listener/pkg/types/event.go#L8), which is then send to the KCP using a secured mTLS connection. To establish a secured mTLS connection from the SKR to the KCP, a secret which holds a TLS certificate will be deployed by the Lifecycle Manager on each SKR as well.
```
Comment (mmitoraj) [inline README.md]:
```suggestion
Runtime Watcher's main part is a validation webhook. The validation webhook's watch mechanism is re-used in a Kyma cluster. Lifecycle Manager configures the webhook using Watcher CRs. The validation webhook has a deployment attached, which is the receiver of the validation requests. The deployment converts the validation requests into [WatchEvents](https://github.com/kyma-project/runtime-watcher/blob/de040bddeba1a7875e3a0e626db4634134971022/listener/pkg/types/event.go#L8), which are then sent to KCP using a secured mTLS connection. To establish the secured mTLS connection from a Kyma cluster to KCP, Lifecycle Manager deploys a Secret with a TLS certificate in each Kyma cluster.
```
You wrote that Runtime Watcher consists of multiple parts but actually, you write only about the validation webhook and what it needed. I would rephrase.

File: README.md [hunk 6]
```diff
@@ -3,19 +3,26 @@
...
+
+
+### [Listener package](https://github.com/kyma-project/runtime-watcher/tree/main/listener)
```
Comment (mmitoraj) [inline README.md]:
```suggestion
### Listener package
```
Let's move the link from the header to the paragraph/

File: README.md [hunk 7]
```diff
@@ -3,19 +3,26 @@
...
+The Listener pkg is provided to simplify setting up an endpoint for an operator (residing inside KCP), which wants to receive                                                                                                                                                                                      Watcher Events send by the Runtime Watcher to the KCP.
```
Comment (mmitoraj) [inline README.md]:
```suggestion
The [Listener package](https://github.com/kyma-project/runtime-watcher/tree/main/listener) simplifies setting up an endpoint for an operator residing in KCP, which should receive WatchEvents sent by Runtime Watcher to KCP.
```
Do I get it right that the Watcher Evenst are WatchEvents described ### Runtime Watcher?

File: docs/04-10-architecture.md [hunk 8]
```diff
@@ -3,37 +3,37 @@
...
 
 The workflow of Runtime Watcher uses the Watcher CR, Runtime Watcher and listener package.
+The following diagram presents Runtime Watcher's workflow.
+
+![Runtime Watcher architecture](./assets/runtime_watcher_architecture_simplified.svg)
 
 ### Watcher CR
 
-The Watcher CR configures Kyma Control Plane (KCP) setup and Runtime Watcher on Kyma runtimes. You can have multiple Watcher CRs configured.
+The Watcher CR has two main purposes: Configuring the KCP landscape and configuring the Runtime Watcher on Kyma Clusters.
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
The Watcher CR's main purpose is to:
- configure the KCP landscape,
- configure Runtime Watcher in Kyma clusters.
```

File: docs/04-10-architecture.md [hunk 9]
```diff
@@ -3,37 +3,37 @@
...
 
-### Runtime Watcher
+To configure the Virtual Services on KCP, the KLM will use the `spec.gateway` and `spec.serviceinfo` field, and the `operator.kyma-project.io/managed-by` label from each Watcher CR. Each Watcher CR will own exact one Virtual Service. It would also be possible to have one big VirtualService including all Watcher CR configuration, but for simplicity it was decided against. If the Watcher configurations changes, or the Watcher CR has been deleted, the KLM will configure the corresponding Virtual Service or delete it as well. This whole mechanism is implemented in its own reconcile loop, which can be found [here](https://github.com/kyma-project/lifecycle-manager/blob/4cb423780633afe7805d26d624c22a6f51943492/controllers/watcher_controller.go#L74).
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
To configure Virtual Services in KCP, Lifecycle Manager uses the `spec.gateway` and `spec.serviceinfo` fields, and the `operator.kyma-project.io/managed-by` label from each Watcher CR. Each Watcher CR owns one Virtual Service. If Runtime Watcher's configuration changes or the Watcher CR is deleted, Lifecycle Manager re-configures the corresponding Virtual Service or deletes it as well. This mechanism is implemented in its own [reconcile loop](https://github.com/kyma-project/lifecycle-manager/blob/4cb423780633afe7805d26d624c22a6f51943492/controllers/watcher_controller.go#L74).
```

File: docs/04-10-architecture.md [hunk 10]
```diff
@@ -3,37 +3,37 @@
...
 
-Runtime Watcher consists of a validation webhook, configured by Watcher CRs, and its deployment. The validation webhook is a resource that watches configured resources and sends validation requests to the attached deployment. The deployment sends requests to KCP.
+The configuration of the Runtime Watcher happens during the reconciliation of a Kyma CR. This means, during each reconciliation of a Kyma CR all Watcher CRs will be fetched and a `ValidationWebhookConfiguration` will be created using the `spec.LabelsToWatch`, `spec.ResourceToWatch` and `spec.Field` from all Watcher CRs. One webhook in the configuration for each Watcher CR. The resulting Validation Webhook Configuration then will be applied to the Kyma cluster as a part of the Runtime Watcher.
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
Runtime Watcher is configured during the reconciliation of a Kyma CR. This means, during each reconciliation of a Kyma CR all Watcher CRs are fetched, and `ValidationWebhookConfiguration` using the `spec.LabelsToWatch`, `spec.ResourceToWatch`, and `spec.Field` from all Watcher CRs is created. The configuration includes one webhook per each Watcher CR. The validation webhook configuration is applied to the Kyma cluster as a part of Runtime Watcher.
```

File: docs/04-10-architecture.md [hunk 11]
```diff
@@ -3,37 +3,37 @@
...
 
-#### Subject Alternative Name (SAN) pinning
+### Runtime Watcher
 
-SAN pinning is an example of a validation function given to the listener pacakage. At least one SAN of the request certificate needs to match the domain specified in the Kyma CR. For that reason, the certificate of an incoming request to Gateway needs to be forwarded to Lifecycle Manager.
+Runtime Watcher consists of a `ValidationWebhookConfiguration`, configured by Watcher CRs, an attached deployment and a secret holding a TLS certificate. The validation webhook is a resource that watches configured resources and sends validation requests to the attached deployment. Instead of validating the received requests, the deployment converts the validation requests into a [WatchEvent](https://github.com/kyma-project/runtime-watcher/blob/de040bddeba1a7875e3a0e626db4634134971022/listener/pkg/types/event.go#L8), which is then send to the KCP using a secured mTLS connection. To establish a secured mTLS connection from the SKR to the KCP, it uses the TLS certificate stored inside a secret. How this secret is being created is [described in this section](###certificates).
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
Runtime Watcher consists of `ValidationWebhookConfiguration` configured by Watcher CRs, an attached deployment, and a Secret holding a TLS certificate. The validation webhook is a resource that watches configured resources and sends validation requests to the attached deployment. Instead of validating the received requests, the deployment converts the validation requests into a [WatchEvent](https://github.com/kyma-project/runtime-watcher/blob/de040bddeba1a7875e3a0e626db4634134971022/listener/pkg/types/event.go#L8), which is sent to KCP using a secured mTLS connection. To establish a secured mTLS connection from a Kyma cluster to KCP, it uses the TLS certificate stored inside a Secret. Go to [certificates](###certificates), to see how this Secret is created.
```
Are the validation requests converted into one WatchEvent or multiple WatchEvents?

File: docs/04-10-architecture.md [hunk 12]
```diff
@@ -3,37 +3,37 @@
...
 
-> **NOTE:** The listener package is temporarily part of the Runtime Watcher repository but it will be moved to the Lifecycle Manager repository.
+Runtime Watcher is configured and deployed on a Kyma cluster in the Kyma reconciliation loop.
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
Runtime Watcher is configured and deployed in a Kyma cluster in the Kyma reconciliation loop.
```

File: docs/04-10-architecture.md [hunk 13]
```diff
@@ -3,37 +3,37 @@
...
 
-## Workflow
+### Listener package
 
-The diagram presents Runtime Watcher's workflow.
+The Listener package is designed to streamline the process of establishing an endpoint for an operator (located within KCP). This operator intends to receive Watcher Events that are transmitted from the Runtime Watcher to the KCP. When calling the `RegisterListenerComponent` function, it returns you a runnable Listener, which should be added to your reconile-manager, and a channel. Example use in the KLM can be found [here](https://github.com/kyma-project/lifecycle-manager/blob/24d21bb642ceaf9dadffe7732bf7c3f70c085ffb/controllers/manifest_controller.go#L43-L50). The channel is then the source the operator should mainly deal with, i.e. it can fetch the incoming watch events from the channel and requeue the corresponding resource in the reconcile loop. Furthermore, it is possible to provide a validation function to the `RegisterListenerComponent` which can be used to already filter out not needed request before processing them further. An example for a verify-function is the SAN pinning, described in the next section.
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
The Listener package is designed to streamline the process of establishing an endpoint for an operator located in KCP. This operator intends to receive Watcher Events that are transmitted from Runtime Watcher to KCP. When calling the `RegisterListenerComponent` function, it returns you a runnable listener, which should be added to your reconile-manager, and a channel. See this [example of how the listener package is used in Lifecycle Manager](https://github.com/kyma-project/lifecycle-manager/blob/24d21bb642ceaf9dadffe7732bf7c3f70c085ffb/controllers/manifest_controller.go#L43-L50). The channel becomes the source for the operator. For example, the operator can fetch the incoming WatchEvents from the channel and requeue the corresponding resource in the reconcile loop. Furthermore, it is possible to provide a validation function to the `RegisterListenerComponent` which can be used to filter out not needed requests before processing them further. An example of the validation function is SAN pinning.
```

File: docs/04-10-architecture.md [hunk 14]
```diff
@@ -3,37 +3,37 @@
...
 
-![Runtime Watcher architecture](./assets/runtime_watcher_architecture_simplified.svg)
+To learn how to setup the listener, have a look [here](./Listener.md)
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
To learn how to set up a listener, read the [Listener document](./Listener.md).
```

File: docs/04-10-architecture.md [hunk 15]
```diff
@@ -3,37 +3,37 @@
...
 
-KCP environment includes multiple Watcher CRs that are reconciled by Lifecycle Manager. It means that Lifecycle Manager configures VirtualServices on KCP. <!--TBD: update the diagram to multiply the VirtualService-->
+#### Subject Alternative Name (SAN) pinning
 
-Istio Gateway is static and the Lifecycle Manager service is attached to the Lifecycle Manager deployment. VirtualServices are configured dynamically by Watcher CRs.
+[SAN pinning](https://github.com/kyma-project/lifecycle-manager/blob/c1e06b7b973aca17cc715b6a4660b76f4e7b9e29/pkg/security/san_pinning.go#L55) is an example implementation of a validation function given to the listener package, and is used in the KLM. The given `Verify` function checks if the certificate of the incoming watcher event request has at least one matching SAN with the domain of the corresponding Kyma CR. The domain of a Kyma CR is saved in an annotation which is called `skr-domain`.
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
[SAN pinning](https://github.com/kyma-project/lifecycle-manager/blob/c1e06b7b973aca17cc715b6a4660b76f4e7b9e29/pkg/security/san_pinning.go#L55) is an example implementation of a validation function given to the listener package. SAN pinning is used in Lifecycle Manager. The validation function checks if the certificate of the incoming WatchEvent request has at least one matching SAN with the domain of the corresponding Kyma CR. The domain of a Kyma CR is saved in an annotation called `skr-domain`.
```

File: docs/04-10-architecture.md [hunk 16]
```diff
@@ -3,37 +3,37 @@
...
 
-Runtime Watcher and certificates <!--TBD: or certificate Secret--> are installed in every Kyma reconciliation in a Kyma runtime.
+### Certificates
+Since it is required to have a mTLS connection from the Kyma cluster to the KCP Gateway, signed TLS certificates are needed for the KCP Gateway as well as for each Kyma cluster. Considering the fact that we do not want to rely on third parties in our infrastructure, we are building our own `Public Key Infrastructure`(PKI) by [bootstraping a selfsigned CA issuer](https://cert-manager.io/docs/configuration/selfsigned/#bootstrapping-ca-issuers). To build this, we are going to leverage the function of the [Cert-Manager](https://cert-manager.io/).
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
Since it is required to have an mTLS connection from the Kyma cluster to the KCP Gateway, signed TLS certificates are needed for both the KCP Gateway and each Kyma cluster. In order not to rely on third parties in the infrastructure, we build the `Public Key Infrastructure`(PKI) by [bootstraping a self-signed CA issuer](https://cert-manager.io/docs/configuration/selfsigned/#bootstrapping-ca-issuers). The PKI uses the function of the [Cert-Manager](https://cert-manager.io/). 
```

File: docs/04-10-architecture.md [hunk 17]
```diff
@@ -3,37 +3,37 @@
...
 
-Certificate Secret includes CA certificate, TLS certificate, and TLS key saved as a Secret and stored in KCP. All of those  are copied to the corresponding Kyma cluster during reconciliation. Lifecycle Manager creates the certificates using [cert-manager](https://github.com/cert-manager/cert-manager) and its self-signed feature. The solution requires Cluster Issuer and Issuer in the `istio-system` Namespace. Cluster Issuer issues a RootCACert. Issuer creates and signs all Kyma certificates using the CA certificate.
+In each Kyma reconciliation loop, the KLM will create or update a [Certificate CR](https://cert-manager.io/docs/concepts/certificate/) for the Kyma CR. This certificate CR will then be signed by a deployed [Issuer](https://cert-manager.io/docs/concepts/issuer/#supported-issuers), which will request the Cert-Manager to create a signed certificate. This certificate will then be stored in a secret on KCP and copied over to the corresponding Kyma cluster when the Runtime Watcher is being deployed. The secret includes the CA certificate, a TLS certificate, and a TLS key.
```
Comment (mmitoraj) [inline docs/04-10-architecture.md]:
```suggestion
In each Kyma reconciliation loop, Lifecycle Manager creates or updates a [Certificate CR](https://cert-manager.io/docs/concepts/certificate/) for the Kyma CR. The Certificate CR is signed by a deployed [Issuer](https://cert-manager.io/docs/concepts/issuer/#supported-issuers), which requests the Cert-Manager to create a signed certificate. This certificate is stored in a Secret in KCP and copied over to the corresponding Kyma cluster when Runtime Watcher is deployed. The Secret includes the CA certificate, a TLS certificate, and a TLS key.
```

File: docs/05-10-api.md [hunk 18]
```diff
@@ -1,25 +1,62 @@
 # Watcher CR
 
-The Watcher CR configures Kyma Control Plane (KCP) setup and Runtime Watcher on Kyma runtimes. Read the document to see the the parameters the CR consists of. **spec.labelsToWatch** Configures the validation webhook configuration in a Kyma runtime (SKR)
+The Watcher CR configures Kyma Control Plane (KCP) setup and Runtime Watcher on Kyma runtimes. Read the document to see the the parameters the CR consists of.
```
Comment (mmitoraj) [inline docs/05-10-api.md]: Add the link, please.

File: docs/05-10-api.md [hunk 19]
```diff
@@ -1,25 +1,62 @@
...
 
-## **spec.labelsToWatch**
+## Runtime-Watcher Configuration
```
Comment (mmitoraj) [inline docs/05-10-api.md]:
```suggestion
## Runtime-Watcher configuration
```

File: docs/05-10-api.md [hunk 20]
```diff
@@ -1,25 +1,62 @@
...
+    version: "*"
+    resource: kymas
+```
 
-## **spec.field**
+### **spec.labelsToWatch**
+> **NOTE:** The resources you're watching need to have the `operator.kyma-project.io/watched-by` label. The label also informs where to route requests to.
 
-<!--TBD-->
+The **spec.labelsToWatch** attribute is used to be able to filter the specified GVK in **spec.resourceToWatch**. For example, if the specified GVK is a `secret`, then it is useful to filter them by a specific label, otherwise the Runtime Watcher would send an event to the KCP for every CRUD event of any secret in the Kyma cluster.
```
Comment (mmitoraj) [inline docs/05-10-api.md]:
```suggestion
The **spec.labelsToWatch** attribute is used to be able to filter the specified Group/Version/Kind (GVK) in **spec.resourceToWatch**. For example, if the specified GVK is `secrets`, then it is useful to filter them by a specific label, otherwise, the Runtime Watcher would send an event to the KCP for every Create/Read/Update/Delete (CRUD) event of any Secret in the Kyma cluster.
```

File: docs/05-10-api.md [hunk 21]
```diff
@@ -1,25 +1,62 @@
...
+    group: ""
+    version: "v1"
+    resource: secrets
+```
 
-<!--TBD-->
 
-## **spec.gateway**
+### **spec.field**
+Can have eather the value `spec` or `status` to define if the Runtime-Watcher should send an event to the KCP if the spec of the specified GVK changes or when the status changes.
```
Comment (mmitoraj) [inline docs/05-10-api.md]:
```suggestion
It uses either the `spec` or `status` value to define if Runtime Watcher should send an event to KCP if the spec of the specified GVK or the status changes.
```

File: docs/05-10-api.md [hunk 22]
```diff
@@ -1,25 +1,62 @@
...
 
-<!--TBD-->
+## KCP Configuration
```
Comment (mmitoraj) [inline docs/05-10-api.md]:
```suggestion
## KCP configuration
```

File: docs/05-10-api.md [hunk 23]
```diff
@@ -1,25 +1,62 @@
...
+
+### **spec.serviceInfo**
+Specifies the name and namespace, as well as the port, the incoming events should be routed to.
```
Comment (mmitoraj) [inline docs/05-10-api.md]:
```suggestion
Specifies the name and Namespace, as well as the port, the incoming events should be routed to.
```

**Review by mmitoraj — COMMENTED** [review]
In some sentences, you were writing about "a Watcher event", and sometimes about "a WatchEvent". Did you refer to the same thing or are those two separate?

File: docs/04-10-architecture.md (hunk 11 above)
Comment (jeremyharisch) [reply docs/04-10-architecture.md]: One request = One Event

**Review by mmitoraj — APPROVED** [review]
